packages/sdk/src/protocols/networkStats.ts

```diff
@@ -0,0 +1,101 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Middleware } from '.';
+import { Message } from '..';
+import { ExportedPromise } from '../utils/exportedPromise';
+import filterEmpty from '../utils/filterEmpty';
+import validateJsonFieldName from '../utils/validateJsonFieldName';
+
+export interface NetworkStatsFrame {
+	messageCount: number;
+	trafficIn: number;
+	trafficOut: number;
+}
+
+export interface NetworkStatsReport {
+	/**
+	 * The average incoming bandwidth of this app over the last 1/5/30 seconds, in KB/s. This is roughly equivalent
+	 * to the bandwidth sent by a single steady-state client over the interval, though does not correlate exactly.
+	 * Only MRE internal traffic is counted, not general HTTP requests (static file hosting, etc.).
+	 */
+	networkBandwidthIn: [number, number, number];
+	/**
+	 * The average outgoing bandwidth of this app over the last 1/5/30 seconds, in KB/s. This is roughly equivalent
+	 * to the bandwidth sent by the busiest client over the interval, though does not correlate exactly.
+	 * Only MRE internal traffic is counted, not general HTTP requests (static file hosting, etc.).
+	 */
+	networkBandwidthOut: [number, number, number];
+	/**
+	 * The number of messages sent and received by this app in the last 1/5/30 seconds. A high number might indicate
+	 * that clients are wasting CPU cycles serializing and deserializing messages.
+	 */
+	networkMessageCount: [number, number, number];
+}
+
+/**
+ * @hidden
+ */
+export class NetworkStats implements Middleware {
+	private buffer: NetworkStatsFrame[] = [];
+	private active: NetworkStatsFrame = {
+		messageCount: 0,
+		trafficIn: 0,
+		trafficOut: 0
+	};
+
+	constructor() {
+		this.beforeSend = this.beforeSend.bind(this);
+		this.beforeRecv = this.beforeRecv.bind(this);
+		setInterval(() => this.cycleStatFrames(), 1000);
+	}
+
+	/** @private */
+	public beforeSend(message: Message, promise?: ExportedPromise): Message {
+		const data = new Buffer(JSON.stringify(message, (key, value) => {
```

Comment:
Benefit to putting `NetworkStats` on `Connection`: You won't have to double-serialize the message to get the size.

Reply:
Actually that's not true. The Pipe connection class deals with objects and not serialized data, so it would still have to manually serialize.

Reply:
oh you're right. that is more efficient, but annoying in this instance.

---
In reply to: [325397097](https://github.com/microsoft/mixed-reality-extension-sdk/pull/445#discussion_r325397097) [](ancestors = 325397097)